00_Base/src/util/middleware/ResponseValidationMiddleware.ts

```diff
@@ -0,0 +1,90 @@
+import {
+  getMetadataArgsStorage,
+  InternalServerError,
+  KoaMiddlewareInterface,
+  Middleware,
+} from 'routing-controllers';
+import { plainToClass } from '../util';
+import { Service } from 'typedi';
+import { Context } from 'vm';
+import { validate } from 'class-validator';
+import { validatedResponseParam } from '../decorators/ValidatedResponseSchema';
+import {
+  OcpiResponse,
+  OcpiResponseStatusCode,
+  OcpiRoutePrefix,
+} from '../../index';
+import { ResponseValidationError } from '../../exception/ResponseValidationError';
+
+@Middleware({ type: 'before' })
+@Service()
+export class ResponseValidationMiddleware implements KoaMiddlewareInterface {
+  async use(
+    context: Context,
+    next: (err?: any) => Promise<any>,
+    ...args: any[]
+  ): Promise<any> {
+    await next();
+
+    const route = this.findRoute(context);
+    if (route) {
+      const { target, method } = route;
+      const responseClass = Reflect.getMetadata(
+        validatedResponseParam,
+        target.prototype,
+        method,
+      );
+
+      if (responseClass && context.body) {
+        const dto: OcpiResponse<any> = plainToClass(
+          responseClass,
+          context.body,
+        );
+        if (dto.status_code === OcpiResponseStatusCode.GenericSuccessCode) {
+          const errors = await validate(dto);
+          if (errors.length > 0) {
+            const errorMessages = errors
+              .map((error) => error.toString())
+              .join('');
+            throw new ResponseValidationError(errorMessages);
+          }
+        } else {
+          // todo handle OcpiErrorResponse and OcpiEmptyResponse
+        }
+      }
+    } else {
+      throw new InternalServerError(
+        'Could not find matching route? This should not happen',
```

Comment:
Very minor, but I'd just update the punctuation to:
`Could not find matching route. This should not happen.`